python/dgl/nn/pytorch/conv/dgnconv.py

```diff
@@ -0,0 +1,200 @@
+
+        Note that using directional aggregation requires the LaplacianPE transform on the input
+        graph for eigenvectors computation (the PE size must be >= k above).
+    scalers: list of str
+        List of scaler function names, selected from:
+
+        * ``identity``: no scaling
+
+        * ``amplification``: multiply the aggregated message by :math:`\log(d+1)/\delta`,
+        where :math:`d` is the in-degree of the node.
+
+        * ``attenuation``: multiply the aggregated message by :math:`\delta/\log(d+1)`
+    delta: float
+        The degree-related normalization factor computed over the training set, used by scalers
+        for normalization. :math:`E[\log(d+1)]`, where :math:`d` is the degree for each node
+        in the training set.
+    dropout: float, optional
+        The dropout ratio. Default: 0.0.
+    num_towers: int, optional
+        The number of towers used. Default: 1. Note that in_size and out_size must be divisible
+        by num_towers.
+    edge_feat_size: int, optional
+        The edge feature size. Default: 0.
+    residual : bool, optional
+        The bool flag that determines whether to add a residual connection for the
+        output. Default: True. If in_size and out_size of the DGN conv layer are not
+        the same, this flag will be set as False forcibly.
+
+    Example
+    -------
+    >>> import dgl
+    >>> import torch as th
+    >>> from dgl.nn import DGNConv
+    >>> from dgl import LaplacianPE
+    >>>
+    >>> transform = LaplacianPE(k=3)
+    >>> g = dgl.graph(([0,1,2,3,2,5], [1,2,3,4,0,3]))
+    >>> g = transform(g)
+    >>> feat = th.ones(6, 10)
+    >>> conv = DGNConv(10, 10, ['dir1-av', 'dir1-dx', 'sum'], ['identity', 'amplification'], 2.5)
+    >>> ret = conv(g, feat)
+    """
+    def __init__(self, in_size, out_size, aggregators, scalers, delta,
+        dropout=0., num_towers=1, edge_feat_size=0, residual=True):
+        aggregators = [AGGREGATORS[aggr] for aggr in aggregators]
+        scalers = [SCALERS[scale] for scale in scalers]
+
+        self.in_size = in_size
+        self.out_size = out_size
+        assert in_size % num_towers == 0, 'in_size must be divisible by num_towers'
+        assert out_size % num_towers == 0, 'out_size must be divisible by num_towers'
+        self.tower_in_size = in_size // num_towers
+        self.tower_out_size = out_size // num_towers
+        self.edge_feat_size = edge_feat_size
+        self.residual = residual
+        if self.in_size != self.out_size:
+            self.residual = False
+
+        self.towers = nn.ModuleList([
+            DGNConvTower(
+                self.tower_in_size, self.tower_out_size,
+                aggregators, scalers, delta,
+                dropout=dropout, edge_feat_size=edge_feat_size
+            ) for _ in range(num_towers)
+        ])
+
+        self.mixing_layer = nn.Sequential(
+            nn.Linear(out_size, out_size),
+            nn.LeakyReLU()
+        )
```

Comment:
You need to define a forward function that directly takes computed Laplacian eigenvectors.

Reply:
That may affect the inheritance from PNA a lot, requiring overloaded forward functions for both DGN and DGNTower. I think it is unnecessary. I prefer the current way of carrying eigenvectors in the graph with a key "eig". How do you think of it?

Reply:
That may affect the inheritance from PNA a lot, requiring overloaded forward functions for both DGN and DGNTower. I think it is unnecessary. I prefer the current way of carrying eigenvectors in the graph with a key "eig". How do you think of it?